dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/trackedentity/DefaultTrackedEntityChangeLogService.java

```diff
@@ -64,7 +65,57 @@ public class DefaultTrackedEntityChangeLogService implements TrackedEntityChange
 
   private final JdbcTrackedEntityChangeLogStore jdbcTrackedEntityChangeLogStore;
 
+  private final TrackedEntityAttributeValueChangeLogStore attributeValueChangeLogStore;
+
+  @Override
+  @Transactional
+  public void addTrackedEntityAttributeValueChangeLog(
+      TrackedEntityAttributeValueChangeLog attributeValueChangeLog) {
+    attributeValueChangeLogStore.addTrackedEntityAttributeValueChangeLog(attributeValueChangeLog);
+  }
+
+  @Override
+  @Transactional(readOnly = true)
+  public List<TrackedEntityAttributeValueChangeLog> getTrackedEntityAttributeValueChangeLogs(
+      TrackedEntityAttributeValueChangeLogQueryParams params) {
+    return aclFilter(attributeValueChangeLogStore.getTrackedEntityAttributeValueChangeLogs(params));
+  }
+
+  private List<TrackedEntityAttributeValueChangeLog> aclFilter(
+      List<TrackedEntityAttributeValueChangeLog> attributeValueChangeLogs) {
+    // Fetch all the Tracked Entity Attributes this user has access
+    // to (only store UIDs). Not a very efficient solution, but at the
+    // moment
+    // we do not have ACL API to check TE attributes.
+
+    Set<String> allUserReadableTrackedEntityAttributes =
+        trackedEntityAttributeService
+            .getAllUserReadableTrackedEntityAttributes(CurrentUserUtil.getCurrentUserDetails())
+            .stream()
+            .map(IdentifiableObject::getUid)
+            .collect(Collectors.toSet());
+
+    return attributeValueChangeLogs.stream()
```

Comment:
what if `allUserReadableTrackedEntityAttributes` is empty? can we return early ?

Reply:
Maybe.
I am just moving some code in this PR, I would prefer not to touch anything at the moment